### ISSUE TITLE
[TS Migration] Improve form validation return type

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#35318]
Coming from https://github.com/Expensify/App/pull/32992#pullrequestreview-1846781855...

### Problem

Currently, the object returned by `validate` is just a `Record<string, string>`, but we know that the return type should be a map of `formInputID: translationKey`, so we're missing an opportunity to be precise and expressive with our type definitions. Some more specific problems with this implementation are:

- It's easy to put whatever string you want in the return type (for example, I was mistakenly using `translate(translationKey)`, when all I needed was `translationKey`). 
- If you use a constant for a form inputID (which is a good practice), then you'll get a type error referencing it in the return type of validate and have to cast it to a normal string (with `as string`).

### Solution
Implement the fixes @blazejkustra laid out [here](https://github.com/Expensify/App/pull/32992#issuecomment-1914296224).

**Comment by blazejkustra** [comment]
I would love to work on this one 😄

**Comment by roryabraham** [comment]
This is kind of blocking my work on https://github.com/Expensify/App/pull/34925/files, and despite lots of effort trying to make progress on it, I'm unfortunately kind of stuck.

I've tried something like this:

```ts
const FORM_IDS = {
    ADD_DEBIT_CARD: 'AddDebitCardForm',
} as const;

// TODO: how can we enforce that all the values of FORM_IDS are included as keys in this,
// without losing type information
const INPUT_IDS = {
    [FORM_IDS.ADD_DEBIT_CARD]: {
        SETUP_COMPLETE: 'setupComplete',
    },
} as const;
```

I've tried this:

```
const INPUT_IDS: Record<ValueOf<typeof FORM_IDS>, Record<string, string>> = {
    [FORM_IDS.ADD_DEBIT_CARD]: {
        SETUP_COMPLETE: 'setupComplete',
    },
} as const;
```

But that doesn't enforce that _all_ values from `FORM_IDS` are included as keys in `INPUT_IDS`. It also makes us lose type information instead of `SETUP_COMPLETE: 'setupComplete'`, we just have `Record<string, string>`.

Maybe there's a `type-fest` utility that can help here? `Exact`, perhaps? We essentially want a mapped type as a constant, but I wasn't able to figure out how to accomplish that.

Then, I got stuck on the next part as well. I was attempting to create a type from `FORM_IDS` and `INPUT_IDS` such that:

- every form has a base set of properties
- every form defines types for all the inputIDs it defined in `INPUT_IDS`

I tried this:

```
type FormData = {
    [FORM_IDS.ADD_DEBIT_CARD]: {
        // TS1170: A computed property name in a type literal must refer to an expression whose type is a literal type or a 'unique symbol' type
        [INPUT_IDS[FORM_IDS.ADD_DEBIT_CARD].SETUP_COMPLETE]: boolean,
    },
};
```

This doesn't work, and again we'd have a similar struggle ensuring that all forms and inputIDs are represented.

**Comment by roryabraham** [comment]
I also ran into some trouble adding `RadioButtons` as a `ValidInput`, specifically around the `onPress` prop. Currently `onPress` is defined to receive a single `GestureResponderEvent` argument, but in reality it can receive different treatment depending on the input that it's used with.

I think the root cause for this problem, essentially, is that `BaseInputProps` is incorrect, in that there's nothing to ensure that all of the inputs accept those props as they're defined.

**Comment by roryabraham** [comment]
Another thing I found a bit confusing was that `FormValueType` is defined loosely as `string | boolean | Date`, but I'd think that the value type would depend on which form is being referenced.

For example, if we're talking about the `FormValueType` for `AddDebitCardForm`, it would basically be: `ValueOf<ValueOf<AddDebitCardForm>>`

**Comment by roryabraham** [comment]
Kind of dumping a lot of ideas here, but that's partially because I'm having difficulty wrapping my head around a lot of this, and just trying to make sense of it.

**Comment by blazejkustra** [comment]
> I think the root cause for this problem, essentially, is that BaseInputProps is incorrect, in that there's nothing to ensure that all of the inputs accept those props as they're defined.

Yes BaseInputProps is the root issue, I'll fix it in a follow-up.

> Another thing I found a bit confusing was that FormValueType is defined loosely as string | boolean | Date, but I'd think that the value type would depend on which form is being referenced.

I'll try to fix it 👍

**Comment by blazejkustra** [comment]
> But that doesn't enforce that all values from FORM_IDS are included as keys in INPUT_IDS. It also makes us lose type information instead of SETUP_COMPLETE: 'setupComplete', we just have Record<string, string>. 

> Maybe there's a type-fest utility that can help here? Exact, perhaps? We essentially want a mapped type as a constant, but I wasn't able to figure out how to accomplish that.

When I create INPUT_IDS mapping I'll make sure it requires to provide all FORM_IDS as keys, it may be complicated but worth trying for sure 👍

**Comment by roryabraham** [comment]
Another NAB improvement that would be nice to see is if we could eliminate the need to define every form key and type twice – once for the form and again for the form draft. This is another thing I couldn't figure out how to do without losing type information. I tried something like this:

```ts
const formIDs = {
    ADD_DEBIT_CARD_FORM: 'addDebitCardForm',
    WORKSPACE_SETTINGS_FORM: 'workspaceSettingsForm',
    WORKSPACE_RATE_AND_UNIT_FORM: 'workspaceRateAndUnitForm',
} as const;

type FormIDs = typeof formIDs;

type DraftFormIDs = {
    [Key in keyof FormIDs | `${keyof FormIDs}_DRAFT`]: Key extends keyof FormIDs ? FormIDs[Key] : `${FormIDs[Replace<Key, '_DRAFT', ''>]}Draft`;
};

const formIDsWithDrafts: DraftFormIDs = Object.entries(formIDs).reduce((acc, [key, value]) => {
    acc[key] = value;
    acc[`${key}_DRAFT`] = `${value}Draft`;
}, {});
```

But that doesn't quite work. I did find [this complicated TS code](https://dev.to/harry0000/a-bit-convenient-typescript-type-definitions-for-objectentries-d6g) that provides a better-typed version of `Object.entries`, but it still didn't make it clear how to accomplish what I was trying to.

**Comment by roryabraham** [comment]
Reviewed @blazejkustra's WIP PR

**Comment by blazejkustra** [comment]
PR is ready for C+ review, I tested in on all platforms and finished reviewer checklist. cc @roryabraham @fabioh8010

**Comment by abdulrahuman5196** [comment]
@roryabraham Could you kindly assign me as C+ here?

**Comment by Julesssss** [comment]
Assigning @abdulrahuman5196 based on [this comment](https://expensify.slack.com/archives/C02NK2DQWUX/p1708607683163729)

**Comment by roryabraham** [comment]
flagging as new feature to handle C+ payment

**Comment by MitchExpensify** [comment]
Reminder set to pay tomorrow

**Comment by abdulrahuman5196** [comment]
No BZ checklist since we improved the TS representation of Form component

cc: @MitchExpensify

**Comment by MitchExpensify** [comment]
Offer here @abdulrahuman5196 https://www.upwork.com/nx/wm/offer/101083068

**Comment by abdulrahuman5196** [comment]
@MitchExpensify accepted the offer
